plutus-contract/src/Wallet/Emulator/ChainIndex/Index.hs

```diff
@@ -0,0 +1,108 @@
+{-# LANGUAGE DeriveGeneric         #-}
+{-# LANGUAGE DerivingStrategies    #-}
+{-# LANGUAGE DerivingVia           #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NamedFieldPuns        #-}
+
+-- | This module implements an index of transactions
+--   by slot and addresses
+module Wallet.Emulator.ChainIndex.Index where
+
+import           Data.FingerTree   (FingerTree, Measured (..), (|>))
+import qualified Data.FingerTree   as FingerTree
+import           Data.Foldable     (foldl', toList)
+import           Data.Map.Strict   (Map)
+import qualified Data.Map.Strict   as Map
+import           Data.Maybe        (fromMaybe)
+import           Data.Semigroup    (Max (..))
+import           GHC.Generics      (Generic)
+
+import           Ledger.Address    (Address)
+import           Ledger.AddressMap (AddressMap)
+import qualified Ledger.AddressMap as AM
+import           Ledger.Slot       (Slot)
+import           Ledger.Tx         (Tx)
+import           Ledger.TxId       (TxId)
+
+-- | The slot in which a transaction was added to the chain.
+newtype TxSlot = TxSlot
+    { unTxSlot :: Maybe (Max Slot)
+    } deriving stock (Eq, Show, Generic)
+      deriving newtype (Semigroup, Monoid)
+
+-- | A transaction with extra information, for the chain index.
+data ChainIndexItem = ChainIndexItem
+    { ciSlot :: !Slot -- ^ The slot in which the transaction was added to the chain
+    , ciTx   :: !Tx -- ^ The transaction
+    , ciTxId :: !TxId -- ^ Hash of the transaction
+    } deriving stock (Eq, Show)
+
+instance Measured TxSlot ChainIndexItem where
+    measure ChainIndexItem{ciSlot} =
+        TxSlot
+            { unTxSlot = Just (Max ciSlot)
+            }
+
+-- | A list of transactions at an address, sorted by slot number
```

Comment:
"Transaction *at* an address" sounds a bit weird and inconsistent with other terminology. Maybe we could say a transaction that "touches" an address?